FAQ.md

```diff
@@ -0,0 +1,112 @@
+* [Unable to update checked out version: fatal: reference is not a tree](#unable-to-update-checked-out-version-fatal-reference-is-not-a-tree)
+
+## What is a direct or transitive dependency?
+* Direct dependencies are dependencies that are imported by your project.
+* Transitive dependencies are the dependencies of your dependencies. Necessary to compile but are not directly used by your code.
+
+## Why is it `dep ensure` instead of `dep install`?
+
+> Yeah, we went round and round on names. [A lot](https://gist.github.com/jessfraz/315db91b272441f510e81e449f675a8b).
+>
+>  The idea of "ensure" is roughly, "ensure that all my local states - code tree, manifest, lock, and vendor - are in sync with each other." When arguments are passed, it becomes "ensure this argument is satisfied, along with synchronization between all my local states."
+>
+>  We opted for this approach because we came to the conclusion that allowing the tool to perform partial work/exit in intermediate states ended up creating a tool that had more commands, had far more possible valid exit and input states, and was generally full of footguns. In this approach, the user has most of the same ultimate control, but exercises it differently (by modifying the code/manifest and re-running dep ensure).
+-[@sdboyer in #371](https://github.com/golang/dep/issues/371#issuecomment-293246832)
+
+## Does `dep` replace `go get`?
+
+> I would say that dep doesn't replace go get, but they both can do similar things. Here's how I use them:
+>
+> `go get`: I want to download the source code for a go project so that I can work on it myself, or to install a tool. This clones the repo under GOPATH for all to use.
+>
+> `dep ensure`: I have imported a new dependency in my code and want to download the dependency so I can start using it. My workflow is "add the import to the code, and then run dep ensure so that the manifest/lock/vendor are updated". This clones the repo under my project's vendor directory, and remembers the revision used so that everyone who works on my project is guaranteed to be using the same version of dependencies.
+-[@carolynvs in #376](https://github.com/golang/dep/issues/376#issuecomment-293964655)
+
+> The long term vision is a sane, overall-consistent go tool. My general take is that `go get`
+> is for people consuming Go code, and dep-family commands are for people developing it.
+-[@sdboyer in #376](https://github.com/golang/dep/issues/376#issuecomment-294045873)
+
+## Why did `dep ensure -update` not update package X?
+
+
+* Is package X a direct dependency? [#385](https://github.com/golang/dep/issues/385)
+
+    Constraints given in a project's manifest are only applied if the
+    dependent project is actually imported. Transitive dependencies (dependencies
+    of your imports) are only updated when the revision in the lockfile no
+    longer meets the constraints of your direct dependencies.
```

Comment:
Sorry I don't quite understand what you mean by "you can (or at least should be able to) mark any dependency, transitive or direct, for updating". Can you help me identity which line(s) are missing the mark? 😊 

I thought that problem from #385 (and #302 come to think of it) was that because X was not a direct dependency (not imported), running `dep ensure -update` did not cause X to be updated, even though X is a dependency in the manifest? So the resolution is either:

* Be okay with it not being updated during `dep ensure -update` unless forced by the constraints of your direct dependencies.
* Make X as an override or required package in the manifest.

Though the best solution is to fix #302 and when dep detects this situation, it stops and prints a message to the user on how to fix it.